frontend/javascripts/oxalis/view/right-border-tabs/segments_tab/segments_view.tsx

```diff
@@ -768,6 +781,365 @@ class SegmentsView extends React.Component<Props, State> {
+                  // Satisfy TS
+                  return;
+                }
+                this.handleChangeMeshVisibility(
+                  this.props.visibleSegmentationLayer.name,
+                  groupId,
+                  !this.state.areSegmentsInGroupVisible[groupId],
+                );
+                this.handleSegmentDropdownMenuVisibility(groupId, false);
+              }}
+            >
+              <i className="fas fa-dice-d20" />
+              {this.state.areSegmentsInGroupVisible[groupId] ? "Hide" : "Show"} all meshes
+            </div>
+          ),
+        }
+      : null;
+  };
+
+  setGroupColor(groupId: number, color: Vector3) {
+    const { visibleSegmentationLayer } = this.props;
+    if (visibleSegmentationLayer == null) return;
+    const segmentGroupToChangeColor = this.getSegmentsOfGroup(groupId);
+    if (segmentGroupToChangeColor == null) return;
+
+    const actionArray = segmentGroupToChangeColor.map((segment) =>
+      updateSegmentAction(segment.id, { color: color }, visibleSegmentationLayer.name),
+    );
+
+    Store.dispatch(batchUpdateGroupsAndSegmentsAction(actionArray));
+  }
+
+  handleRefreshMeshes = (groupId: number) => {
+    const { visibleSegmentationLayer } = this.props;
+    if (visibleSegmentationLayer == null) return;
+    const segmentGroup = this.getSegmentsOfGroup(groupId);
+    if (segmentGroup == null) return;
+
+    segmentGroup.forEach((segment) => {
+      if (
+        Store.getState().localSegmentationData[visibleSegmentationLayer.name].isosurfaces[
+          segment.id
+        ] != null
+      ) {
+        Store.dispatch(refreshIsosurfaceAction(visibleSegmentationLayer.name, segment.id));
+      }
+    });
+  };
+
+  handleRemoveMeshes = (groupId: number) => {
+    const { visibleSegmentationLayer } = this.props;
+    if (visibleSegmentationLayer == null) return;
+    const segmentGroup = this.getSegmentsOfGroup(groupId);
+    if (segmentGroup == null) return;
+
+    segmentGroup.forEach((segment) => {
+      if (
+        Store.getState().localSegmentationData[visibleSegmentationLayer.name].isosurfaces[
+          segment.id
+        ] != null
+      ) {
+        Store.dispatch(removeIsosurfaceAction(visibleSegmentationLayer.name, segment.id));
+      }
+    });
+  };
+
+  handleChangeMeshVisibility = (layerName: string, groupId: number, isVisible: boolean) => {
+    const { visibleSegmentationLayer } = this.props;
+    if (visibleSegmentationLayer == null) return;
+    const segmentGroup = this.getSegmentsOfGroup(groupId);
+    if (segmentGroup == null) return;
+
+    segmentGroup.forEach((segment) => {
+      if (Store.getState().localSegmentationData[layerName].isosurfaces[segment.id] != null) {
+        Store.dispatch(updateIsosurfaceVisibilityAction(layerName, segment.id, isVisible));
+      }
+    });
+
+    const copyOfState = this.state.areSegmentsInGroupVisible;
+    copyOfState[groupId] = isVisible;
```

Comment:
this is not a copy. instead, you are only binding the variable `copyOfState` to the object `areSegmentsInGroupVisible` and mutate it then. a "copy" via variable assignment is only done for primitive variable types.

what you probably want to do is something like this (not tested):

```suggestion
    const newSegmentsInGroupVisible = {
       ...this.state.areSegmentsInGroupVisible,
       [groupId]: isVisible
    }
```

Reply:
oh right, I think it's something you told me about in the very beginning. thanks for the reminder!